common/test/java/com/couchbase/lite/BaseReplicatorTest.kt

```diff
@@ -15,28 +15,81 @@
 //
 package com.couchbase.lite
 
+import com.couchbase.lite.internal.replicator.ReplicationStatusChange
 import com.couchbase.lite.internal.utils.Fn
 import com.couchbase.lite.internal.utils.Report
-import com.couchbase.lite.mock.TestReplicatorChangeListener
 import org.junit.After
 import org.junit.Assert
 import org.junit.Before
 import java.net.URI
 import java.security.cert.Certificate
 import java.security.cert.X509Certificate
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Executor
 import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+internal class ListenerAwaiter(private val token: ListenerToken) {
+    companion object {
+        private val stoppedStates = setOf(
+            ReplicatorActivityLevel.STOPPED,
+            ReplicatorActivityLevel.OFFLINE,
+            ReplicatorActivityLevel.IDLE
+        )
+    }
+
+    private val err = AtomicReference<Throwable?>(null)
+    private val latch = CountDownLatch(1)
+
+    val error: Throwable?
+        get() = err.get()
+
+    fun changed(change: ReplicationStatusChange) {
+        val status = change.status
+        status.error?.let { err.compareAndSet(null, it) }
+        val level = status.activityLevel
+        val e = status.error
+        Report.log(e, "Test replicator state change: $level")
+
+        if (e != null) err.compareAndSet(null, e)
+
+        if (stoppedStates.contains(level)) latch.countDown()
```

Comment:
Noted : if there is a test that add docs in different rate (e.g. keep adding docs in parallels), the IDLE status cannot be trust to determine that the replicator is finished replicating as it could be BUSY -> IDLE -> BUSY - IDLE. I recently found this issue in one of the LiteCore's flaky tests.